crypto/encrypt/encrypt.go

```diff
@@ -21,38 +21,40 @@ import (
 	"crypto/rand"
 )
 
-func Bytes(bytes []byte) ([]byte, []byte, error) {
-	// Should be 16 bytes (AES-128), 24 bytes (AES-192) or 32 bytes (AES-256)
-	key := make([]byte, 32)
-	_, err := rand.Read(key)
-	if err != nil {
-		return nil, nil, err
+func Bytes(key, bytes []byte) ([]byte, []byte, []byte, error) {
+	// Key should be 16 bytes (AES-128), 24 bytes (AES-192) or 32 bytes (AES-256)
+	if key == nil {
+		key = make([]byte, 32)
+		_, err := rand.Read(key)
+		if err != nil {
+			return nil, nil, nil, err
+		}
 	}
 
 	// Generate a new aes cipher using the key above
 	block, err := aes.NewCipher(key)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// gcm or Galois/Counter Mode, is a mode of operation
 	// for symmetric key cryptographic block ciphers
 	// - https://en.wikipedia.org/wiki/Galois/Counter_Mode
-	gcm, err := cipher.NewGCM(block)
+	aesGCM, err := cipher.NewGCM(block)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// Create a new byte array the size of the nonce,
 	// populate with a cryptographically secure random sequence
-	nonce := make([]byte, gcm.NonceSize())
+	nonce := make([]byte, aesGCM.NonceSize())
 	_, err = rand.Read(nonce)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, err
 	}
 
 	// Encrypt and authenticate plaintext
-	ciphertext := gcm.Seal(nonce, nonce, bytes, nil)
+	ciphertext := aesGCM.Seal(nonce, nonce, bytes, nil)
 
-	return ciphertext, key, nil
+	return ciphertext, key, nonce, nil
```

Comment:
Actually, now I am wondering if we should define and return a struct here, which would also include `tag`, `tagSize`, etc. @jeremyletang @rickymarcon WDYT? I am not sure we should return more than 3 items as that becomes a bit difficult to follow for the consumer. Thoughts?

Reply:
For example - we would `return struct, nil` here.